commit/src/domain.rs

```diff
@@ -0,0 +1,364 @@
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_field::{
+    batch_multiplicative_inverse, cyclic_subgroup_coset_known_order, ExtensionField, Field,
+    TwoAdicField,
+};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_util::{log2_ceil_usize, log2_strict_usize};
+
+use crate::two_adic_decompose::decompose;
+
+pub struct LagrangeSelectors<T> {
+    pub is_first_row: T,
+    pub is_last_row: T,
+    pub is_transition: T,
+    pub inv_zeroifier: T,
+}
+
+pub trait PolynomialDomain: Copy {
+    type Val: Field;
+
+    fn size(&self) -> usize;
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Ext;
+
+    // There are many choices for this, but we must pick a canonical one
+    // for both prover/verifier determinism and LDE caching.
+    fn create_disjoint_domain(&self, min_size: usize) -> Self;
+
+    // decompose evaluations until the domain size is less than max_size
+    // and return the resulting domain
+    fn decompose(
+        &self,
+        evals: RowMajorMatrix<Self::Val>,
+        max_size: usize,
+    ) -> RowMajorMatrix<Self::Val>;
+    fn decomposed_domain(&self, max_size: usize) -> Self;
+    fn decomposed_point<Ext: ExtensionField<Self::Val>>(&self, point: Ext, max_size: usize) -> Ext;
+    fn decomposed_width_factor(&self, max_size: usize) -> usize;
+
```

Comment:
I have these methods on `Domain`
```rust
pub trait PolynomialDomain {
     fn decompose(
         &self,
         evals: RowMajorMatrix<Self::Val>,
         max_size: usize,
     ) -> RowMajorMatrix<Self::Val>;
     fn decomposed_domain(&self, max_size: usize) -> Self;
     fn decomposed_point<Ext: ExtensionField<Self::Val>>(&self, point: Ext, max_size: usize) -> Ext;
     fn decomposed_width_factor(&self, max_size: usize) -> usize;
```

for decomposing polynomials that are larger than your PCS's degree bound.

There's some redundancy. The main reason for all this is to (for example, for two-odic) ask the PCS to open at `zeta^2` instead of `zeta`. Some redundancy could be eliminated by putting the `decomposed_bits` into `TwoAdicDomain` itself, so it keeps track of how many times to square opened points. Mathematically, that kind of oversteps what a "domain" is, but maybe it would make sense from a code perspective.

(We need `decompose` to do the actual decomposition, `decomposed_domain` to pass to pcs to commit/verify on the decomposed domain, `decomposed_point` to open/verify at `zeta^2` instead of `zeta`, and `decomposed_width_factor` to verify proof shape). And `recompose` to put the openings back together at `zeta`.